File: agent-harness/cmd/harness/tokens.go
package main

import (
	"encoding/json"
	"log"
	"strings"
)

// tokenMarkerPrefix tags a log line carrying the run's cumulative token usage as
// a compact JSON object, e.g. `BANDOLIER_TOKENS={"input_tokens":4,...}`. Like the
// PR_URL/ISSUE_URL markers, it rides the pod log (and the persisted transcript)
// so Bandolier can surface and persist the count without a side channel. The
// server greps the most recent occurrence, so emitting the running total each
// turn keeps the readout live for interactive sessions.
const tokenMarkerPrefix = "BANDOLIER_TOKENS="

// tokenUsage is the per-run token accounting reported by the agent CLIs. The
// fields mirror Anthropic's usage shape (the claude CLI's stream-json `result`
// event); cache fields are zero for providers that don't report them. Kept flat
// (no nested objects) so the server can extract it with a single-brace regex.
type tokenUsage struct {
	InputTokens              int `json:"input_tokens"`
	OutputTokens             int `json:"output_tokens"`
	CacheReadInputTokens     int `json:"cache_read_input_tokens"`
	CacheCreationInputTokens int `json:"cache_creation_input_tokens"`
}

// empty reports whether no tokens were recorded — used to skip emitting a marker
// for an event that carried no usage.
func (u tokenUsage) empty() bool {
	return u.InputTokens == 0 &&
		u.OutputTokens == 0 &&
		u.CacheReadInputTokens == 0 &&
		u.CacheCreationInputTokens == 0
}

// add accumulates another turn's usage into this one. Interactive sessions drive
// many turns through one long-lived process; summing keeps a session-wide total.
func (u *tokenUsage) add(o tokenUsage) {
	u.InputTokens += o.InputTokens
	u.OutputTokens += o.OutputTokens
	u.CacheReadInputTokens += o.CacheReadInputTokens
	u.CacheCreationInputTokens += o.CacheCreationInputTokens
}

// latestTokenMarker scans a transcript for the most recent token marker and
// returns its JSON payload (the substring after the prefix, e.g.
// `{"input_tokens":4,...}`). Returns "" when none is present. The last marker
// wins: one-shot runs emit once, interactive runs emit a growing total per turn,
// so the final occurrence is the run's cumulative figure.
func latestTokenMarker(transcript string) string {
	idx := strings.LastIndex(transcript, tokenMarkerPrefix)
	if idx < 0 {
		return ""
	}
	rest := transcript[idx+len(tokenMarkerPrefix):]
	if nl := strings.IndexByte(rest, '\n'); nl >= 0 {
		rest = rest[:nl]
	}
	return strings.TrimSpace(rest)
}

// logTokenUsage emits the token marker into the pod log / transcript. No-op for
// empty usage so a provider that reports nothing leaves no misleading "0" marker.
func logTokenUsage(u tokenUsage) {
	if u.empty() {
		return
	}
	b, err := json.Marshal(u)
	if err != nil {
		return
	}
	log.Printf("[harness] %s%s", tokenMarkerPrefix, b)
}

File: src/lib/tokens.ts
// Token-usage helpers shared by the harness ingest path, the agents router, and
// the dashboard. The harness emits a `BANDOLIER_TOKENS={json}` marker line into
// the pod log / transcript (see agent-harness/cmd/harness/tokens.go); these
// helpers parse that marker, sum a run's total, and format it for display.

/** A run's token accounting. Mirrors the harness's flat tokenUsage JSON. */
export interface TokenUsage {
  inputTokens: number;
  outputTokens: number;
  cacheReadInputTokens: number;
  cacheCreationInputTokens: number;
}

// Marker prefix the harness logs the usage JSON behind. This crosses the
// process boundary (the Go harness emits it, this parses it), so its value is
// pinned in wire-contract.json and asserted by both test suites — see
// src/lib/wire-contract.test.ts.
export const TOKEN_MARKER = "BANDOLIER_TOKENS=";

function toInt(v: unknown): number {
  return typeof v === "number" && Number.isFinite(v) && v > 0
    ? Math.floor(v)
    : 0;
}

/**
 * Parses the harness's token marker payload (the JSON after `BANDOLIER_TOKENS=`)
 * into a TokenUsage. Tolerant of missing/garbage fields — anything unparseable
 * yields null so callers can fall back to "no usage" rather than a wrong number.
 */
export function parseTokenMarkerPayload(json: string): TokenUsage | null {
  let raw: unknown;
  try {
    raw = JSON.parse(json);
  } catch {
    return null;
  }
  if (typeof raw !== "object" || raw === null) return null;
  const o = raw as Record<string, unknown>;
  const usage: TokenUsage = {
    inputTokens: toInt(o.input_tokens),
    outputTokens: toInt(o.output_tokens),
    cacheReadInputTokens: toInt(o.cache_read_input_tokens),
    cacheCreationInputTokens: toInt(o.cache_creation_input_tokens),
  };
  return usage;
}

/**
 * Scans a transcript / log blob for the most recent token marker and returns the
 * parsed usage. The last marker wins: one-shot runs emit once, interactive runs
 * emit a growing total per turn, so the final occurrence is the run's cumulative
 * figure. Returns null when no (valid) marker is present.
 */
export function parseTokenUsageFromLogs(logs: string): TokenUsage | null {
  const idx = logs.lastIndexOf(TOKEN_MARKER);
  if (idx < 0) return null;
  let rest = logs.slice(idx + TOKEN_MARKER.length);
  const nl = rest.indexOf("\n");
  if (nl >= 0) rest = rest.slice(0, nl);
  return parseTokenMarkerPayload(rest.trim());
}

/** The grand total of every token category — what the readouts display. */
export function totalTokens(u: TokenUsage): number {
  return (
    u.inputTokens +
    u.outputTokens +
    u.cacheReadInputTokens +
    u.cacheCreationInputTokens
  );
}

/**
 * Compact human-readable token count for the dashboard: 1234 → "1.2K",
 * 1_500_000 → "1.5M". Exact below 1000. Used in the task row, log modal, and
 * interactive session card.
 */
export function formatTokens(n: number): string {
  if (n < 1000) return String(n);
  if (n < 1_000_000) {
    const k = n / 1000;
    return `${k >= 100 ? Math.round(k) : k.toFixed(1).replace(/\.0$/, "")}K`;
  }
  const m = n / 1_000_000;
  return `${m >= 100 ? Math.round(m) : m.toFixed(1).replace(/\.0$/, "")}M`;
}
